packages/components/docs/immer.md

```diff
@@ -0,0 +1,235 @@
+When one of these objects is mutated within a producer, its prototype is preserved between copies.
+
+Now this class is ready to be used with Immer. Let's try again using `produce`:
+
+```ts
+import { produce } from 'immer';
+
+// without using produce the instance is still mutable
+let circle = new Circle(5);
+circle.radius = 20; // radius now 20.
+
+// with produce the instance is now immutable
+circle = produce(new Circle(5), () => {});
+circle.radius = 10; // non-strict mode: fails silently. Radius still 5.
+circle.radius = 10; // strict mode: Run-time error: Cannot assign to read only property 'radius' of object '#<Circle>'
+```
+
+Instances of this class declared via `produce` can only be mutated via `produce`. When an instance is passed through
+Immer's `produce` function it will [freeze](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze)
+the object (when `immer.setAutoFreeze(true)`). Depending on strict mode any attempts to explicitly modify the object
+will either fail to modify or throw a run-time error.
+
+Now, let's actually make an update to the immutable instance with Immer:
+
+```ts
+import { produce } from 'immer';
+
+// Create the initial instance
+const circle = produce(new Circle(5), () => {});
+
+// Mutate and copy from `produce`
+const newCircle = produce(circle, (draft) => {
+    draft.radius = 10;
+});
+console.log(circle.radius);    // 5
+console.log(newCircle.radius); // 10
+```
+
+The instance is now immutable, a mutated copy can be made via `produce`, and we have run-time safety via `Object.freeze`.
+
+#### Compile-time safety
+
+The main advantage of using TypeScript is that it compiles your code to ensure correctness before you run it. Additionally,
+the run-time safety is great but it still may be something that only occurs in production if say our test coverage misses
+a code path.
+
+To get started, let's first declare all the properties on the `Circle` class as [`readonly`](https://www.typescriptlang.org/docs/handbook/classes.html#readonly-modifier).
+
+```ts
+import { immerable } from 'immer';
+
+class Circle {
+    [immerable] = true;
+
+    readonly radius: number;
+
+    constructor(r: number) {
+        this.radius = r;
+    }
+}
+```
+
+The `radius` property is now read-only so if we attempt to modify it directly we receive an error:
+
+```ts
+import { produce } from 'immer';
+
+let circle = new Circle(5);
+circle.radius = 5; // Error: TS2540: Cannot assign to 'radius' because it is a read-only property.
+
+// Same for the produced version
+let circle = produce(new Circle(5), () => {});
+circle.radius = 5; // Error: TS2540: Cannot assign to 'radius' because it is a read-only property.
+```
+
+This gives us compile-time safety against invalid writes. The next feature we can use is the `Draft` utility from
+Immer. To quote the docs:
+
+> For curried reducers, the type is inferred from the first argument of recipe function, so make sure to type it.
```

Comment:
Curried reducers?  Recipe function?  This is quite foreign at this point in this document, and I'm not even sure that sentence parses well.

Reply:
agreed, doesn't seem needed here

Reply:
We should probably link directly to the Immer docs:

https://immerjs.github.io/immer/docs/curried-produce

And maybe even to an explanation of currying.

Reply:
I've snipped the quote to the most relevant part for this context:

> The Draft utility type can be used if the state argument type is immutable.

This aligns nicely with the example directly following it.